### ISSUE TITLE
Generate JS tuple from F# tuple

### ISSUE THREAD

**Issue by sydsutton** [fable-compiler/Fable#3546]
### Description

I was wondering if there are any configuration options for me to purposefully produce a JS (or TS) tuple from an F# tuple instead of a JS array from F# tuple? I'm working on an F# front-end library, and need a tuple for this particular implementation. Thanks!

Function:
`static member inline mergeClasses (classNames: 'T): string = import "mergeClasses" FluentUIv9`

Usage: 
`prop.className (Fui.mergeClasses (styles.backgroundColor, styles.textColor, styles.borderRadius))`

Fable compilation:
`createObj(ofArray([["className", mergeClasses([styles.backgroundColor, styles.textColor, styles.borderRadius])]`

### Expected and actual results

The code works as it should if Fable were to compile it into:

`createObj(ofArray([["className", mergeClasses(styles.backgroundColor, styles.textColor, styles.borderRadius)]` (no array)

### Related information

* Fable version: `4.1.4`
* Operating system: Windows 11

**Comment by ncave** [comment]
@sydsutton Not supported yet at the moment, no.
It does makes sense to add optional support for JS tuples and records, I'll tag this as a feature request.

In the mean time, is it possible to wrap the tuple arrays in `Tuple.from` ?
I know it's probably a bad work-around, so only as a temporary solution.

**Comment by sydsutton** [comment]
@ncave Thanks for the quick response! I'll give it a shot. What namespace would `Tuple.from` be under?

**Comment by sydsutton** [comment]
No, I am not sure about browser support, but I have been successfully using tuples for functions in my project for a while now just by using partial application. Thanks for the response, btw.

**Comment by ncave** [comment]
@sydsutton I guess my question was, what JavaScript engine have you used those new #Tuples with?
Somehow I can't find any references to actual support, besides a few articles about a preview spec.

**Comment by MangelMaxime** [comment]
It seems like `Tuple` in JavaScript are still in [Stage 2: Draft](https://github.com/tc39/proposal-record-tuple) and [Can I use](https://caniuse.com/?search=tuple) doesn't report any results.

So I don't think tuple are yet widely spread meaning that their implementation is subject to change. We can't add support for it in Fable directly, because we don't know if this going to be implemented or not in the future. And also, if the specification change then we would need to introduce breaking changes.

I summarised Ncave utilities + another example in this [REPL](https://fable.io/repl/#?code=PYBwpgdgBAygngZwC5gLYChSSgMQIYBGANmAHQDCwATmJuNPsWZTaQFIICSEKVo66APSCoATWABXKAGNJRACZQQRPNLBQkACwCWCKKmDyJJKBIjywVKHElXgAdwaESFamTYx9h42ARCRAGbUMsAQCLpIeBBq6AZGJh5QALwCUGlQJEhQ2hBEOeqoeADWYAAqEspg3NpZAB7JuM7MbuxcPJagpGg1HACitSBWABS1AJRQAEQAxAAkAAwT6OkZYFmFJeWVOHyoUPVJjUyurBzcvJ3dSH0Dw2OTmy4BO0Pzo4voANoAPL1UeAhgAB8AF10Eg4OAoKVfFkUkt0sg8EhtNJ9GgCJZsrl8miqABzMDkFQIAF6IbSYkIAByeFQvigAC4oAByUrjA7aVAgahZCZ0-GEym+CaTHBECSQJAAVU4ADcAJzvZaI5GoumoDFWHJ5CAFSwEon-UlQIaygCMjJZpQANFBZQAmS2s9lQeHLNLQ5CkfkGoVkjze4plCokapZU1m20O0ajVIIyKqtEazHanE+wVG+kRp02u2OpmsqMAZhzLrd7s9SG9+ozJKzAfWwcqYZN5qj9uLMYEmSgyDgJgO5aGExo8gmtomBGACgmsfQlerAsNdb0fZI85hi99mf9MEDGxDYG2wF2a7AsYX6eXxqGDaDDyqEBqvfBJHGwigVOAWQQEho2QCDQdD0exqCKPxLxra96z3RsHxbYdR3HSYpxnGMoA-GAwDQPQkGAKACV1P4UCA9RZCoGhpCySQkBACQkAESClz9E0RzAMcJ1QscL03K8WMQ9jkMnacOMmMwLCoHUwFnIA&html=DwCwLgtgNgfAUAAlAUwIYBN4O0iyyoLhgAOAtMgI4CuAlgG4C8A5AMID2AdmMt2QCoBPEsmYIAxlx7cWPAB5gA9OGgBuCSFQAnAM75G1MADMyADmZYcwHeK20SYBDq3jGAIgD6HgBIB5AMr8XopQtABGOopaaOJgAHQkWuzo1LG0XHEQtJxxAFY6bjDAijZ2DpbY1rb2js6unj4BQR4h4ZHRqLFk6OwQCUkpaRlZOfmFxaU1FUihnADWCNFQ7jpgglDIOiDI+G5E0UbuXn6BwaERiuI6kWHUUBComdlxVwUIitPAswtLK2sbWx2YD2IAOR0apxa50ir0UqCgUCeOVeew+cGK2wwWGAYWSggkUFQ13cqBI5Ek3FQ2WQWkKiCs6AYCFo6HcyHutC2ZFJJD24kJxLc7KyXJ540UjPo8GKuPQgmxykgsCAA&css=HQYwhgTgJsCWDOBaMA3MAXSACLBvAUDjgLZgAeiA7rFOgBYBcWAzAKwAMADmQNyFGkIAc1gA7JmACu6APZ8AvoVCQYCZGkwQsy6IlikhAUzz8cUBJwA2YAJ5MAZpcO9TWAFaT46WPZuIQMqLohkFMICHBEHxEWGCWsEKiesHE8GERhlGunGBQ5qJCTABMrNxY7OUKQA) link.

```fs
open System
open Fable.Core
open Fable.Core.JsInterop

// You could place this module under your own Fable.Core.JS modules
// for consistance
module JS =

    let inline makeTupleInit x = Fable.Core.JsInterop.emitJsExpr (x) "#$0"
    let makeTupleFrom x = Fable.Core.JsInterop.emitJsExpr (x) "Tuple.from($0)"

[<Erase>]
type Test =

    static member inline mergeClasses (classNames : 'T) = import "mergeClasses" "FluentUIv9"

    static member inline mergeClasses (v1 : 'T, v2 : 'T) = 
        Test.mergeClasses (JS.makeTupleInit (v1, v2))

    static member inline mergeClasses (v1 : 'T, v2 : 'T, v3 : 'T) = 
        Test.mergeClasses (JS.makeTupleInit (v1, v2, v3))

let style = 
    ("red", "bold")

Test.mergeClasses style
Test.mergeClasses (JS.makeTupleFrom style)
Test.mergeClasses (JS.makeTupleInit style) // Not sure if this works
Test.mergeClasses (JS.makeTupleInit ("red", "bold")) // Seems to generate the correct output

Test.mergeClasses ("red", "bold")
Test.mergeClasses ("red", "bold", "underline")
```

Another solution to add support for Tuple in your project, could be by using a Fable compiler plugin. Like Feliz does for React, but it needs more works and it is not well documented yet.

**Comment by sydsutton** [comment]
@MangelMaxime @ncave Great, thank you guys for the responses. I really appreciate the help.

**Comment by ken-okabe** [comment]
I have encountered the same issue because [VanJS](https://vanjs.org/) heavily depends on n-ary functions api that I need to use.

My workaround is to create a bridge.js and export the hooked functions like this

```js
import van from 'vanjs-core'

// unary function ([a,b,c,...]) in F#
// -> n-ary function (a,b,c,...) in VanJS
let n =
    f => array =>
        f(...array);

export let tags =
    new Proxy(van.tags, {
        get: (target, property) => {
            return n(target[String(property)]);
        }
    });

export let add = n(van.add);

```

then imports from *.fs

```fsharp
open Fable.Core.JsInterop

let tags: obj
       = importMember "../ts/tags_add"

let add: List<obj> -> Element
       = importMember "../ts/tags_add"

```

**Comment by ken-okabe** [comment]
I'm sorry, but probably, I misunderstood your question.
As already mentioned here, JS/TS currently does not have tuple data structure, so in theory, "Generate JS tuple" is impossible.

In JS, `(a, b, c)` is not a value, but `f(a, b, c)` means n-ary function that takes multiple arguments. 

My workaround is just convert 

`f [a, b, c]`  in F#
to
`f (a, b, c)` in JS

**Comment by MangelMaxime** [comment]
I am going to close this issue for now.

Tuple proposition has not moved since last time, once it is widely supported we can revisit
